### ISSUE TITLE
Question on permissions

### ISSUE THREAD

**Issue by IzzySoft** [rRemix/APlayer#227]
My scanner recently got additional checks implemented, and on your latest release reported:

```
! repo/remix.myplayer_16201.apk declares sensitive permission(s):
  android.permission.MANAGE_EXTERNAL_STORAGE android.permission.REQUEST_INSTALL_PACKAGES
  android.permission.READ_EXTERNAL_STORAGE android.permission.SYSTEM_ALERT_WINDOW
  android.permission.READ_MEDIA_AUDIO android.permission.READ_MEDIA_IMAGES
! repo/remix.myplayer_16201.apk contains signature block blobs: 0x504b4453 (DEPENDENCY_INFO_BLOCK; GOOGLE)
```

Now I could easily align `READ_EXTERNAL_STORAGE` and `READ_MEDIA_AUDIO` being needed to access the audio files to play. But could you please clarify what the other permissions are needed for? Thanks in advance!

Btw, the `DEPENDENCY_INFO_BLOCK` can easily be avoided:

```gradle
android {
    dependenciesInfo {
        // Disables dependency metadata when building APKs.
        includeInApk = false
        // Disables dependency metadata when building Android App Bundles.
        includeInBundle = false
    }
}
```

For some background: that BLOB is supposed to be just a binary representation of your app's dependency tree. But as it's encrypted with a public key belonging to Google, only Google can read it – and nobody else can even verify what it really contains.

**Comment by rRemix** [comment]
**android.permission.MANAGE_EXTERNAL_STORAGE**, such as reading lyric file, blacklist
**android.permission.SYSTEM_ALERT_WINDOW**, desktop lyric
**android.permission.REQUEST_INSTALL_PACKAGES**, in-app upgrade, but it's only available in non-google channel

**Comment by Henry-ZHR** [comment]
`MANAGE_EXTERNAL_STORAGE` is initially for "Manual scan" function (users may want to scan files outside standard directories), and it may also solve some permission-related issues on some Chinese systems. It's not necessary for everyone.

**Comment by IzzySoft** [comment]
> in-app upgrade, but it's only available in non-google channel

It's been reported for the APK used in my repo. And self-updater are not in accordance with the repo's inclusion criteria – as those updates would bypass the checks performed in the repo. May I ask how it is configured? Enabled by default? Opt-in (so with clear consent)? Explaining the source of the update and the implications (such as bypassing repo-checks)?

**Comment by Henry-ZHR** [comment]
`DEPENDENCY_INFO_BLOCK` excluded in b1f990191175ba7fb30a371b17e70017a3d7dec8

**Comment by IzzySoft** [comment]
Wonderful, thanks! So what's with `REQUEST_INSTALL_PACKAGES` now (see my last comment)?

**Comment by Henry-ZHR** [comment]
I suggest adding a new `flavor` which disables the self-updater (and maybe the non-free dependencies?)

Anyway, @rRemix is the only owner of this repo :)

**Comment by IzzySoft** [comment]
Thanks Henry! OK, so let's see what the owner says :smile:

**Comment by rRemix** [comment]
i'll add it

**Comment by IzzySoft** [comment]
Thanks, @rRemix! Please let me know when that flavor is available. I'd suggest to switch your app to that in my repo then. If it involves a new `applicationId`/`packageName`, is there a "migration path" (e.g. export/import settings)?

**Comment by rRemix** [comment]
@IzzySoft you can use flavor 'noUpdater'

**Comment by Henry-ZHR** [comment]
@rRemix 不太对 我说开flavor是方便你那边直接release加一个 因为我以为他们是直接用你的apk 但是他们好像是自己build 那感觉不如开buildConfigField

@IzzySoft Are you building the app from source?

**Comment by IzzySoft** [comment]
> Are you building the app from source or using the binary from GitHub Release?

I use the APKs provided by their authors – so yes, I'd need the corresponding APK at releases. Thanks!

**Comment by Henry-ZHR** [comment]
@rRemix 那就还是放flavor吧

但是还有些小问题

1. 大小写不能随便动 如果要动源码目录得跟着改

2. 现在命名有点怪 感觉这仨flavor不是同级的 以及可以放另外的dimension？

**Comment by Henry-ZHR** [comment]
@rRemix

1. 其实我也觉得用大小写分隔单词更合理 只是提醒你记得改源码 Linux文件系统一般区分大小写（

2. https://developer.android.com/build/build-variants#filter-variants 可以filter掉google+updater组合？

以及 忘说了 应该其他都算非google渠道 如果在同一个dimension应该不用依赖billingclient？（在新的dimension就没有这个问题

**Comment by Henry-ZHR** [comment]
> @rRemix
> 
>     1. 其实我也觉得用大小写分隔单词更合理 只是提醒你记得改源码目录 Linux文件系统一般区分大小写（
> 
>     2. https://developer.android.com/build/build-variants#filter-variants 可以filter掉google+updater组合？
> 
> 
> 以及 忘说了 应该其他都算非google渠道 如果在同一个dimension应该不用依赖billingclient？（在新的dimension就没有这个问题

**Comment by IzzySoft** [comment]
Today's release still has my scanner report

    android.permission.REQUEST_INSTALL_PACKAGES android.permission.READ_MEDIA_IMAGES

So what did I miss concerning the latter? And will there be an APK for the NoUpdater flavor?

**Comment by Henry-ZHR** [comment]
Sorry, I forgot `android.permission.REQUEST_INSTALL_PACKAGES`. It will be included whether self-updater is enabled or not until #248 is merged.

**Comment by IzzySoft** [comment]
Thanks! So that means after that PR is merged, `android.permission.REQUEST_INSTALL_PACKAGES` will be gone? And what about `android.permission.READ_MEDIA_IMAGES`?

**Comment by Henry-ZHR** [comment]
`READ_MEDIA_IMAGES` **may** be needed to read the cover image of music

We need @rRemix or someone else who is familiar with Android permission system to check whether it's really necessary

**Comment by IzzySoft** [comment]
That would be one way to find out. If you have the chance, you could just compile an APK with that permission commented out and try that on-device (or EMU). AFAIK `READ_MEDIA_IMAGES` refers to photos (DCIM) and maybe screenshots, but not to images embedded in IDv3 or such – but I'm not 100% sure and thus would appreciate if someone could "tell definitely". Thanks for taking care!

**Comment by Henry-ZHR** [comment]
APlayer does read the cover not only directly from ID3 tag but also from MediaStore. For example: <https://github.com/rRemix/APlayer/blob/3a0f4613523be47f7f169b36106e24ee4da6da7e/app/src/main/java/remix/myplayer/bean/mp3/Song.kt#L107>

It should require `READ_MEDIA_AUDIO` only, but I'm not sure.

**Comment by IzzySoft** [comment]
> Sorry, I forgot android.permission.REQUEST_INSTALL_PACKAGES. It will be included whether self-updater is enabled or not until https://github.com/rRemix/APlayer/pull/248 is merged.

That is merged, but the permission is still present.

It's almost 2 years now since I first reported this, that should be sufficient time to solve the problem. It seems like we have to remove the app from our repo, if that cannot be solved.

> you can use flavor 'noUpdater'

There never was an APK from that flavor available here at your releases.

I've now put a note to remove APlayer in 3 weeks from now, unless that problem has been addressed. Apologies if it sounds a bit harsh – but what good is a policy if it's not enforced? And the [IzzyOnDroid App Inclusion Policy](https://izzyondroid.org/docs/general/AppInclusionPolicy/) is clear on this.

@rRemix @Henry-ZHR please address this ASAP.

**Comment by rRemix** [comment]
> > Sorry, I forgot android.permission.REQUEST_INSTALL_PACKAGES. It will be included whether self-updater is enabled or not until https://github.com/rRemix/APlayer/pull/248 is merged.
> 
> That is merged, but the permission is still present.
> 
> It's almost 2 years now since I first reported this, that should be sufficient time to solve the problem. It seems like we have to remove the app from our repo, if that cannot be solved.
> 
> > you can use flavor 'noUpdater'
> 
> There never was an APK from that flavor available here at your releases.
> 
> I've now put a note to remove APlayer in 3 weeks from now, unless that problem has been addressed. Apologies if it sounds a bit harsh – but what good is a policy if it's not enforced? And the [IzzyOnDroid App Inclusion Policy](https://izzyondroid.org/docs/general/AppInclusionPolicy/) is clear on this.
> 
> @rRemix @Henry-ZHR please address this ASAP.

sorry,my bad.i'll add "withoutUpdater" apk in release later.

**Comment by IzzySoft** [comment]
Thanks! Please give me a ping once it's up!

**Comment by rRemix** [comment]
@IzzySoft https://github.com/rRemix/APlayer/releases/download/v2.0.0.0.0/APlayer-v2.0.0.0-nonGoogle-withoutUpdater-release.apk

**Comment by IzzySoft** [comment]
Thanks! Doing a rare thing now at our end, replacing the APK that was already fetched earlier today. Confirming the INSTALL permission being gone. If you now could give me a short hint what `android.permission.READ_MEDIA_IMAGES` is needed for, we have all gaps closed.

**Comment by IzzySoft** [comment]
Btw, I was trying whether I could reproduce the APK (see: [Reproducible Builds](https://izzyondroid.org/about/security/ReproducibleBuilds/)) – very close, the only difference remaining are BuildIDs. Could you let me know which path you build in (e.g. `/home/rremix/aplayer`), and which path your AndroidSDK is installed to (`$ANDROID_HOME`), so I can match those in our recipe? Might also help if you could tell your `cmake` to not use build ids, [see here](https://izzyondroid.org/docs/reproducibleBuilds/RBDevHints/#no-funny-build-time-generated-ids) for examples).

Side note: interesting to see that the `GITHUB_SHA` seems to be unused in this build (that `GOOGLE_PLAY_LICENSE_KEY` is missing in the `noGoogle` flavor, is no surprise).

And by the way: happy new year!

[diff.zip](https://github.com/user-attachments/files/24398615/diff.zip)

**Comment by rRemix** [comment]
> Thanks! Doing a rare thing now at our end, replacing the APK that was already fetched earlier today. Confirming the INSTALL permission being gone. If you now could give me a short hint what `android.permission.READ_MEDIA_IMAGES` is needed for, we have all gaps closed.

First of all, Happy New Year to you too! I've made some changes and performed a quick test. The READ_MEDIA_IMAGES permission is no longer required. Please check the latest APK.
https://github.com/rRemix/APlayer/releases/download/v2.0.0.0.0/APlayer-v2.0.0.0-nonGoogle-withoutUpdater-release.apk

**Comment by rRemix** [comment]
> Btw, I was trying whether I could reproduce the APK (see: [Reproducible Builds](https://izzyondroid.org/about/security/ReproducibleBuilds/)) – very close, the only difference remaining are BuildIDs. Could you let me know which path you build in (e.g. `/home/rremix/aplayer`), and which path your AndroidSDK is installed to (`$ANDROID_HOME`), so I can match those in our recipe? Might also help if you could tell your `cmake` to not use build ids, [see here](https://izzyondroid.org/docs/reproducibleBuilds/RBDevHints/#no-funny-build-time-generated-ids) for examples).
> 
> Side note: interesting to see that the `GITHUB_SHA` seems to be unused in this build (that `GOOGLE_PLAY_LICENSE_KEY` is missing in the `noGoogle` flavor, is no surprise).
> 
> And by the way: happy new year!
> 
> [diff.zip](https://github.com/user-attachments/files/24398615/diff.zip)

Here is the info you requested:

Build Path: /Users/remix/AndroidProjects/APlayer
Android SDK Path ($ANDROID_HOME): /Users/remix/Library/Android/sdk

I've updated CMakeLists.txt to disable build IDs (added "-Wl,--build-id=none"), which should resolve the native library differences.Also, I've removed the unused GITHUB_SHA(https://github.com/rRemix/APlayer/commit/755b2d5092f9dec003a1baf0501e6aa8a6d50761).

**Comment by IzzySoft** [comment]
> The READ_MEDIA_IMAGES permission is no longer required. Please check the latest APK.

Confirmed, thanks! But please, in the future, do **not** replace APK files which were already distributed. We've made an exception replacing the APK yesterday, but that means folks who already had 2.0.0.0.0 installed, won't get this as update: our updater simply will ignore them, as "we already have this release".

Concerning RB: I could not test it, as the commit your APK was built from (2aae4fd05ec9443555d32fe31f89d3c48f45a6d5) does not exist here (so we could not build our APK to compare against) – you probably have not yet pushed it. Please let me know once that commit is available here at your Github repo, so I can try again.

**Comment by rRemix** [comment]
> > The READ_MEDIA_IMAGES permission is no longer required. Please check the latest APK.
> 
> Confirmed, thanks! But please, in the future, do **not** replace APK files which were already distributed. We've made an exception replacing the APK yesterday, but that means folks who already had 2.0.0.0.0 installed, won't get this as update: our updater simply will ignore them, as "we already have this release".
> 
> Concerning RB: I could not test it, as the commit your APK was built from (2aae4fd05ec9443555d32fe31f89d3c48f45a6d5) does not exist here (so we could not build our APK to compare against) – you probably have not yet pushed it. Please let me know once that commit is available here at your Github repo, so I can try again.

Understood, I'll keep that in mind for the next version release . For this current one, I've updated it to be based on the latest commit ( 2aae4fd05ec9443555d32fe31f89d3c48f45a6d5 ).

**Comment by IzzySoft** [comment]
> Understood, I'll keep that in mind for the next version release

Thanks! Will have to wait for the next release then, cannot pull the same one another time.

> For this current one, I've updated it to be based on the latest commit ( 2aae4fd05ec9443555d32fe31f89d3c48f45a6d5 ).

Which is not the commit the tag points to, and not even present here at Github – so no way we can check if it would be reproducible… So this task, too, has to wait for the next release. A ping would be welcome then. Recommended reading on Reproducible Builds: our [RB Hints for Developers](https://izzyondroid.org/docs/reproducibleBuilds/RBDevHints/) 😉

**Comment by rRemix** [comment]
@IzzySoft https://github.com/rRemix/APlayer/releases/tag/v2.0.1.0.0

**Comment by IzzySoft** [comment]
Thanks! But… built from 6262739c4af6601302906e7f518e44e1296aadbb – while the tag points to 67e7f2510f54f1afb165e3b0f4c3d595169a80b1 … Well, I guess built before committing those changes. Let's see if that works out, it's a bit lottery this way. For RB (and also generally), the APK attached to a release should be built from a clean tree at the tagged commit. That newer commit has a lot of changes. I hope you didnt "commit › change things › build › _change more things_ › commit › tag", but omitted the point in italics – else we have no chance to get it RB.

Hm, close, but no cigar. Seems like `BUGLY_APP_ID` should NOT be set for the `nonGoogleWithoutUpdater` variant (my build had it, but yours no longer; not sure if you just forgot to add it to the `local.properties`, or if it's intentionally left out – but if it's in the file, it is used).

Oh and ah and see: Now it is erR Bee! 🥳 So [here comes your welcome wagon](https://floss.social/@IzzyOnDroid/115837907782849446):

<img width="322" height="583" alt="Image" src="https://github.com/user-attachments/assets/c34ad4a1-d00e-402d-86cf-048d87db7ef8" />

Thanks – and congrats! For future releases, please remember to "build from a clean tree at the tagged commit" 😉 And just for interest, here are our collected [RB Hints for Developers](https://izzyondroid.org/docs/reproducibleBuilds/RBDevHints/).

**Comment by IzzySoft** [comment]
BTW, @rRemix: it would be nice if you can decide whether the `BUGLY_APP_ID` is needed/used or not. v2.0.2.0 seems to need it again, so RB failed and required manual adjustments, while v2.0.1.0.0 failed because it was present 🙈 I was able to adjust the recipe to make it succeed again. So just please not toggle it back and forth now 😉 Thanks!

**Comment by rRemix** [comment]
Sorry for the confusion. BUGLY_APP_ID is definitely needed. I accidentally removed it from local.properties in the last version, which caused the inconsistency. It will stay included from now on. @IzzySoft

**Comment by IzzySoft** [comment]
Ah! OK, all good then, thanks!

**Comment by rRemix** [comment]
@IzzySoft 
**Updated flavors**: I've added a **foss** flavor which removes Bugly and Google dependencies (including BUGLY_APP_ID). Please note that the APK name for the FOSS flavor is now **APlayer-vx.x.x.x-foss-release.apk**.

**Comment by IzzySoft** [comment]
Thanks for the heads-up! Can you please give me a ping once a corresponding pre-release is up, so I can prepare everything at our end?

**Comment by rRemix** [comment]
To avoid breaking the built-in updater in older versions of my app (which might grab the wrong APK if mixed), I've moved the **FOSS** builds to a separate release tag.

Please update the fetch rule to target tags and releases with the -foss suffix (e.g., v2.0.4.0-foss ).

- Standard Release: **Release-v2.0.4.0-20401** (Marked as Latest, for standard users, https://github.com/rRemix/APlayer/releases/tag/v2.0.4.0-foss)
- FOSS Release: **Release-v2.0.4.0-20401-foss** (https://github.com/rRemix/APlayer/releases/tag/v2.0.4.0)

Let me know if there are any issues, thanks

**Comment by IzzySoft** [comment]
Thanks a lot! Our scanners are happy with the latest FOSS build. As that's the only APK at the latest release: will it stay the only one there? Asking because that would be good to know for when making which adjustments (we can only update our RB recipes when a release is there that matches the updates). One remaining issue would be that, to support release filtering (and only fetching the `-foss` tags), a pending PR would need to be merged first at rbtlog (which I had no time yet to properly test, hence it's still pending). But as APK names differ as well, that would be no big deal, just the opposite: it would make a good test case (and before the merge, only cause some additional warnings in the logs when the latest tag is NOT the `-foss` one).

**Comment by rRemix** [comment]
Yes, I plan to keep it this way:
1. The **Standard** release will always be marked as "Latest" for general users (APK filenames typically include `-normal`).
2. The **FOSS** builds will continue to reside in their own separate release tags (ending with `-foss`), and those releases will exclusively contain the FOSS APK.

I'm glad to hear this structure works for your scanners and can serve as a test case for your pending PR.  @IzzySoft

**Comment by IzzySoft** [comment]
That means, for a while no RB, as the updater there won't find any matching "latest" releases (which is what the update process there is looking for). Why not putting both APKs on the same release? Their names are distinctive enough, aren't they? That's also what other projects with multiple flavors do.

I cannot yet tell when the new PR'd code will be merged. So if you're choosing the path you described, RBs will need manual updates – not just manually triggered ones, but manual editing of the recipe files, as the updater won't fetch new releases properly. Yeah, would be a good test case – only that I don't know yet when I can run it. Maybe in a week or two, idk – still too much load on my task list here…

**Comment by rRemix** [comment]
Sorry, I misunderstood the sorting order of assets in GitHub releases. I have now consolidated all APKs into a single release.
https://github.com/rRemix/APlayer/releases/tag/v2.0.4.1 @IzzySoft

**Comment by IzzySoft** [comment]
Thanks! That unfortunately does not reproduce. I've opened a new issue for that, see #337 – something goes wrong there with the build…
